x-pack/plugins/ml/public/util/app_state_utils.js

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+
+import { distinctUntilChanged } from 'rxjs/operators';
+
+export function initializeAppState(AppState, stateName, defaultState) {
+  const appState = new AppState();
+  appState.fetch();
+
+  // Store the state to the AppState so that it's
+  // restored on page refresh.
+  if (appState[stateName] === undefined) {
+    appState[stateName] = _.cloneDeep(defaultState) || {};
+    appState.save();
+  }
+
+  // If defaultState is defined, check if the keys of the defaultState
+  // match the one from appState, if not, fall back to the defaultState.
+  // If we didn't do this, the structure of an out-of-date appState
+  // might break some follow up code. Note that this will not catch any
+  // deeper nested inconsistencies.
+  if (typeof defaultState !== 'undefined' && appState[stateName] !== defaultState) {
+    // if defaultState is an object, check if current appState has the same keys.
+    // if it's not an object, check if defaultState and current appState are of the same type.
+    if (
+      (typeof defaultState === 'object' && !_.isEqual(
+        Object.keys(defaultState).sort(),
+        Object.keys(appState[stateName]).sort()
+      )) ||
+      (typeof defaultState !== typeof appState[stateName])
+    ) {
+      appState[stateName] = _.cloneDeep(defaultState);
+      appState.save();
+    }
+  }
+
+  return appState;
+}
+
+export function subscribeAppStateToObservable(AppState, appStateName, o$, $rootScope) {
```

Comment:
rather than passing in `$rootScope` could this take a `refresh` function (or some better name) instead which is passed `$rootScope.$applyAsync`
That way it's not tied to angular.
It would probably need to be wrapped in a `typeof === 'function'` check for safety

Reply:
Good idea, I changed that last argument to just `callback` which gets called on each subscription trigger in 4a3578b.